File: divDigits.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

using namespace cv;
using namespace std;

inline Mat calcHist(const Mat& image);
int whichPattern(const Mat& image);


// obtained through experiment
const int pattern1_thresh = 220;
const int pattern2_thresh = 210;
const int min_connected_len = 15; // minimum length of connected components which will not discard

int main(int argc, char *argv[])
{	
	if(argc != 5) {
		cout << "Usage: breaker <path-to-data> <digit-write-path> <start-index> <end-index>" << endl;
		exit(-1);
	}

	string path_to_data = argv[1];
	string digit_write_path = argv[2];
	int start_index = atoi(argv[3]);
	int end_index = atoi(argv[4]);

	for(int i = start_index; i <= end_index; i++){
		string file = path_to_data + "/" + to_string(i) + ".jpeg";
		Mat srcImage = imread(file, 0);
		
		int pattern = whichPattern(srcImage); // determine pattern of the image 

		Mat threshImg;
		
		int thresh = pattern == 1 ? pattern1_thresh : pattern2_thresh;

		threshold(srcImage, threshImg, thresh, 255, THRESH_BINARY_INV);
		
		Mat labels, stats, centroids;
		int noComp = connectedComponentsWithStats(threshImg, labels, stats, centroids);
		
		Mat binaryImage(labels.size(), CV_8UC1);

		// holds if component will be included to foreground
		vector<bool> includeComp(noComp);
		vector<int> foreComps;
		includeComp[0] = false; // background
	
		for(int i = 1; i < noComp; i++)
			// if size less than <t> pixels dont include
			includeComp[i] = (stats.at<int>(i, CC_STAT_AREA) >= min_connected_len) ? (foreComps.push_back(i), true) : false;
		

		for(int y = 0; y < labels.rows; y++)
			for(int x = 0; x < labels.cols; x++)
			{
				int label = labels.at<int>(y, x);
				if(includeComp[label] == true)
					binaryImage.at<uchar>(y, x) = 255;
				else
					binaryImage.at<uchar>(y, x) = 0;
			}
		
		
		int minCol = 30; // seen that all digits start at 30th column 
						 // no need for additional computation
		

		int maxCol = 0;
		for(int i = 0; i < foreComps.size(); i++){
			int col = stats.at<int>(foreComps[i], CC_STAT_LEFT) + stats.at<int>(foreComps[i], CC_STAT_WIDTH);
			if(col > maxCol && col < 125)
				maxCol = col;
		}

		if(pattern == 2) // apply some offset for pattern2
			if(maxCol <= 111)
				maxCol = 111;
			else if (maxCol < 117)
				maxCol = 117;
		
		int minRow = binaryImage.rows;
		for(int i = 0; i < foreComps.size(); i++){
			int row = stats.at<int>(foreComps[i], CC_STAT_TOP);
			if(row < minRow )
				minRow = row;
		}

		int maxRow = 0;
		for(int i = 0; i < foreComps.size(); i++){
			int row = stats.at<int>(foreComps[i], CC_STAT_TOP) + stats.at<int>(foreComps[i], CC_STAT_HEIGHT);
			if(row > maxRow )
				maxRow = row;
		}

		Rect subReg = Rect(minCol, minRow, maxCol - minCol, maxRow - minRow);
		Mat subImage = Mat(binaryImage,subReg);
		
		vector<Mat> digits;

		Rect subImage1 = Rect(0, 0, subImage.cols / 2, subImage.rows);
		Rect subImage2 = Rect(subImage.cols / 2, 0, subImage.cols / 2, subImage.rows);

		Mat subMat1 = Mat(subImage, subImage1);
		Mat subMat2 = Mat(subImage, subImage2);

		int colIncrement1 = subMat1.cols / 3; 
		int colIncrement2 = subMat2.cols / 3; 
		int col = 0;
		for(int i = 0; i < 2; i++){
			Rect subDigit = Rect(col, 0, colIncrement1, subImage.rows);
			digits.push_back(Mat(subMat1, subDigit));
			col += colIncrement1;
		}
		Rect subDigit = Rect(col, 0, subMat1.cols-col, subImage.rows);
		digits.push_back(Mat(subMat1, subDigit));
		
		col = 0;
		for(int i = 0; i < 2; i++){
			Rect subDigit = Rect(col, 0, colIncrement2, subImage.rows);
			digits.push_back(Mat(subMat2, subDigit));
			col += colIncrement2;
		}
		subDigit = Rect(col, 0, subMat2.cols-col, subImage.rows);
		digits.push_back(Mat(subMat2, subDigit));

		for(int j = 0; j < digits.size(); j++){
			string name =  digit_write_path	+ "/" + to_string(i) + "_" + to_string(j) + ".jpeg";
			imwrite(name, digits[j]);
		}
	}
}

Mat calcHist(const Mat& image)
{

	int histSize = 256;
	float range[] = {0, 256};
	const float* histRange = {range};
	int channels[] = {0};
	Mat imgHist; 

	calcHist(&image, 1, channels, Mat(), imgHist, 1, &histSize, &histRange);

	return imgHist;
}

// finds out whether given image falls under pattern1 or pattern2
int whichPattern(const Mat& image)
{
	Mat hist = calcHist(image);
	
	// pattern1 images has almost more than 600 pixels
	// for 0(pure black) intensity, lets set threshold to 500
	int thresh = 500;
	if(hist.at<float>(0) > thresh)
		return 1; // pattern1
	else
		return 2; // pattern2

	
}

File: divDigits.py
import cv2
import sys
import numpy as np

#obtained through experiment 
pattern1_thresh = 170
pattern2_thresh = 145
min_connected_len = 15

def whichPattern(image):
	hist = cv2.calcHist([image], [0], None, [256], [0, 256])
	# pattern1 images has almost more than 600 pixels
	# for 0(pure black) intensity, lets set threshold to 500
	thresh = 500	
	if(hist[0, 0] > thresh):
		return 1
	else:
		return 2

# Separates the image <file> to six subimages containing each digit
# With an attempt to internally eliminate noise and color inversion
def separate_image(file):
	srcImage = cv2.imread(file, 0)

	pattern = whichPattern(srcImage) # determine the pattern of the image

	thresh = pattern1_thresh if pattern == 1 else pattern2_thresh

	# test if gaussian works
	srcImage = cv2.medianBlur(srcImage, 3) if pattern == 1 else cv2.GaussianBlur(srcImage, (3,3), 0)
	
	#if pattern == 1:
	#	srcImage2 = cv2.GaussianBlur(srcImage, (3,3), 0)

	#cv2.imshow("Median", srcImage)
	
	#global threshImage
	#while cv2.waitKey() != ord('c') and pattern1_thresh != 255:
	ret, threshImage = cv2.threshold(srcImage, thresh, 255, cv2.THRESH_BINARY_INV)
	#cv2.imshow("threshold", threshImage)
	#print(thresh, end='\r', flush=True)
	#thresh += 1
	'''
	if pattern == 1:	
		kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (3,3))
		#threshImage = cv2.morphologyEx(threshImage, cv2.MORPH_DILATE, kernel, 1)
		#cv2.imshow("after dilate", threshImage)
		threshImage = cv2.morphologyEx(threshImage, cv2.MORPH_CLOSE, kernel, 2)
		cv2.imshow("after close", threshImage)
		#kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (2,2))
		#threshImage = cv2.morphologyEx(threshImage, cv2.MORPH_CLOSE, kernel, 1)
		#cv2.imshow("after close", threshImage)
		#threshImage = cv2.morphologyEx(threshImage, cv2.MORPH_ERODE, kernel, 1)
		#cv2.imshow("after erode", threshImage)
	'''	
	numLabel, labelImage, stats, centroids = cv2.connectedComponentsWithStats(threshImage, 8, cv2.CV_32S)
	
	# holds if compenet will be included to foreground
	foreComps = [i for i in range(1, numLabel) if stats[i, cv2.CC_STAT_AREA] >= min_connected_len]
	
	
	minCol = 30; # seen that all digits start at 30th column 
							 # no need for additional computation
	
	binaryImage = np.zeros_like(srcImage)
	labelImage = np.array(labelImage)
	for k in [np.where(labelImage == i) for i in foreComps]:
		binaryImage[k] = 255
	

	array = np.array([stats[i, cv2.CC_STAT_LEFT] + stats[i, cv2.CC_STAT_WIDTH]  for i in foreComps])
	maxCol = max(array[np.where(array < 125)])
		
	# Dont touch
	if pattern == 2:
		if maxCol <= 111:
			axCol = 111
		elif maxCol < 117:
			maxCol = 117

	minRow = min([stats[i, cv2.CC_STAT_TOP] for i in foreComps])
	maxRow = max([stats[i, cv2.CC_STAT_TOP] + stats[i, cv2.CC_STAT_HEIGHT] for i in foreComps])

	#subImage = binaryImage[minRow:maxRow, minCol:maxCol]
	#cv2.imshow("subImage", subImage)
	
	subImage = threshImage[minRow:maxRow, minCol:maxCol]
	#cv2.imshow("subImage", subImage1)

	subImage1 = subImage[:, :int(subImage.shape[1]/2)]
	subImage2 = subImage[:, int(subImage.shape[1]/2):]

	colIncrement1 = subImage1.shape[1] / 3
	colIncrement2 = subImage2.shape[1] / 3
	
	digitList1 = []
	digitList2 = []
	
	col1 = 0
	col2 = 0
	for i in range(2):
		digitList1.append(subImage1[:, int(col1):int(col1+colIncrement1)])
		digitList2.append(subImage2[:, int(col2):int(col2+colIncrement2)])
		col1 += colIncrement1
		col2 += colIncrement2

	digitList1.append(subImage1[:, int(col1):])
	digitList2.append(subImage2[:, int(col2):])
	
	if cv2.waitKey() == ord('q'):
		sys.exit(1)

	return digitList1 + digitList2
  
def divDigits(args):
	if(len(args) != 5):
		print("Usage: breaker <path-to-data> <digit-write-path> <start-index> <end-index>")
		sys.exit(-1)

	path_to_data = args[1];
	digit_write_path = args[2];
	start_index = int(args[3]);
	end_index = int(args[4]);

	for idx in range(start_index, end_index + 1):
		file = path_to_data + "/" + str(idx) + ".jpeg"
		digit_list = separate_image(file)
		print(idx, end='\r', flush=True)
		for j in range(6):
			cv2.imwrite(digit_write_path + '/digit_' + str(idx) + '_' +  str(j) + '.jpeg', digit_list[j])

if __name__ == '__main__':
	divDigits(sys.argv)
